core/embed/rust/src/trezorhal/layout_buf.rs

```diff
@@ -0,0 +1,69 @@
+use core::{
+    marker::PhantomData,
+    mem::{align_of, size_of, MaybeUninit},
+    ptr::NonNull,
+};
+
+pub use super::ffi::c_layout_t;
+
+pub struct LayoutBuffer<T> {
+    layout: NonNull<c_layout_t>,
+    _marker: PhantomData<T>,
+}
+
+impl<T> LayoutBuffer<T> {
+    /// Create a new `LayoutBuffer` from an underlying buffer.
+    ///
+    /// # Safety
+    ///
+    /// Caller must ensure that the buffer can be held mutably for the lifetime
+    /// of the `LayoutBuffer`.
+    pub unsafe fn new(buffer: *mut c_layout_t) -> Self {
+        let layout = unwrap!(NonNull::new(buffer));
+        // SAFETY: the layout is non-null and properly aligned
```

Comment:
IMHO, no one guarantees that `c_layout_t` is properly aligned - since it’s allocated on the stack, it’s currently aligned to 4 bytes. That might be sufficient for all layouts and could always work, but wouldn’t it be safer to split the array using `align_to_mut()` here and use the properly aligned pointer?

Reply:
you're right, I misread NonNull docs and thought that it guarantees alignment which it doesn't.

i believe it doesn't matter with the types that we're using, because `c_layout_t` is just an array of u8 which is always aligned ... and the alignment requirement for `T` is explicitly checked a couple lines down

i'm wondering if we need the whole song-and-dance about `c_layout_t` struct? but that's probably the easiest way to ensure the buffer size across the C-Rust boundary, correct? but if we could just pass a buffer + size from C to Rust, we wouldn't need to worry about changes to `c_layout_t` struct

Reply:
at minimum we can fix the comment
```suggestion
        // SAFETY: c_layout_t is just an array of u8 which is always aligned
```

Reply:
I think the problem is that `c_layout_t` is defined as a struct containing a `uint8_t` array - so the pointer to this array is not necessarily aligned. References to this array might be typically aligned to 4 bytes, but I would not rely on it.

If the type `T` requires a higher alignment (e.g., 8 bytes), the code may panic .

Reply:
how about this? https://github.com/trezor/trezor-firmware/commit/4dbfbac0fba7be34963db43f844c1618525eb4d5

enforcing the alignment via proposed align_to_mut 

(in successive work, i used c_layout_t allocated statically, which was then not aligned to 4 bytes and this paniced as prediced, i went for enforcing alignment via gcc attribute but this seems more robust.

Reply:
Looks good. Maybe we can try to get rid of redundant alignment 

and instead of storing of `layout: NonNull<c_layout_t>,`
store the aligned pointer as `layout: NonNull<MaybeUninit<T>>, 

so that `get_maybe_uninit()` would be simpler

Reply:
one more variant: https://github.com/trezor/trezor-firmware/commit/e74a04c696710d0663df97a8abe70b31cba23cae

IIUC we still need the marker as NonNull is essentially a pointer and we expect ownership of underlying data